feat: add C implementation for `math/base/special/rising-factorial`

DhruvArvindSingh opened the pull request:
Progresses #649  .

## Description

> What is the purpose of this pull request?

This pull request:

-   adds C implementation for [math/base/special/rising-factorial](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/math/base/special/rising-factorial).
```c
double stdlib_base_rising_factorial( const double x, const int32_t n )
```

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   progresses #649  

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

DhruvArvindSingh committed:
feat: add C implementation for math/base/special/rising-factorial

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

DhruvArvindSingh committed:
fix(readme): fixed readme

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

anandkaranubc committed:
Merge branch 'develop' into feat/rising-factorial_develop

Signed-off-by: Karan Anand <119553199+anandkaranubc@users.noreply.github.com>

anandkaranubc commented:
/stdlib merge

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/rising-factorial/src/main.c:
```diff
@@ -0,0 +1,114 @@
+*
+* ## Notice
+*
+* The original C++ code and copyright notice are from the [Boost library]{@link http://www.boost.org/doc/libs/1_64_0/boost/math/special_functions/factorials.hpp}. The implementation has been modified according to project conventions.
```
```suggestion
* The original C++ code and copyright notice are from the [Boost library]{@link http://www.boost.org/doc/libs/1_64_0/boost/math/special_functions/factorials.hpp}. The implementation has been modified according to stdlib conventions.
```

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/rising-factorial/src/main.c:
```diff
@@ -0,0 +1,114 @@
+	double xc;
+	bool inv;
+
+	if ( stdlib_base_is_nan( x ) || !stdlib_base_is_integer( n ) ) {
```
```suggestion
	if ( stdlib_base_is_nan( x ) ) {
```

`n` is already and integer.

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/rising-factorial/src/main.c:
```diff
@@ -0,0 +1,114 @@
+#include "stdlib/math/base/assert/is_integer.h"
+#include "stdlib/math/base/assert/is_nan.h"
+#include "stdlib/math/base/special/gamma_delta_ratio.h"
+#include "stdlib/math/base/special/falling_factorial.h"
```
```suggestion
#include <stdint.h>
```

anandkaranubc replied:
```suggestion
#include <stdbool.h>
```

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/rising-factorial/src/main.c:
```diff
@@ -0,0 +1,114 @@
+	nc = n;
+	xc = x;
+	inv = false;
+	if ( xc < 0.0 ) {
```
```suggestion
		// For `x < 0`, we really have a falling factorial, modulo a possible change of sign. Note that the falling factorial isn't defined for negative `n`, so we'll get rid of that case first:
```

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/rising-factorial/src/main.c:
```diff
@@ -0,0 +1,114 @@
+			nc = -nc;
+			inv = true;
+		}
+		result = ( ( nc & 1 ) ? -1.0 : 1.0 ) * stdlib_base_falling_factorial( -xc, nc );
```
```suggestion
		result = ( (nc&1) ? -1.0 : 1.0 ) * stdlib_base_falling_factorial( -xc, nc );
```

Following the same negative spacing in `main.js`

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/rising-factorial/src/main.c:
```diff
@@ -0,0 +1,114 @@
+	xc = x;
+	inv = false;
+	if ( xc < 0.0 ) {
+		if ( nc < 0.0 ) {
```
```suggestion
		if ( nc < 0 ) {
```

nc is `int32_t`

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/rising-factorial/src/main.c:
```diff
@@ -0,0 +1,114 @@
+		}
+		return 0.0;
+	}
+	if ( ( xc < 1.0 ) && ( xc + nc < 0.0 ) ) {
```
```suggestion
	if ( ( xc < 1.0 ) && ( xc+nc < 0.0 ) ) {
```

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/rising-factorial/src/main.c:
```diff
@@ -0,0 +1,114 @@
+	}
+	if ( ( xc < 1.0 ) && ( xc + nc < 0.0 ) ) {
+		result = stdlib_base_gamma_delta_ratio( 1.0 - xc, -nc );
+		return ( nc & 1 ) ? -result : result;
```
```suggestion
		return ( nc&1 ) ? -result : result;
```

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/rising-factorial/src/main.c:
```diff
@@ -0,0 +1,114 @@
+		return ( nc & 1 ) ? -result : result;
+	}
+
+	// We don't optimize this for small `nc`, because `stdlib_base_gamma_delta_ratio` is already optimized for that use case:
```
```suggestion
	// We don't optimize this for small `n`, because `stdlib_base_gamma_delta_ratio` is already optimized for that use case:
```

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/rising-factorial/src/addon.c:
```diff
@@ -0,0 +1,23 @@
+#include "stdlib/math/base/special/rising_factorial.h"
+#include "stdlib/math/base/napi/binary.h"
+
+// cppcheck-suppress shadowFunction
```
```suggestion
```

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/rising-factorial/lib/native.js:
````diff
@@ -0,0 +1,85 @@
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*
+*
+* ## Notice
+*
+* The original C++ code and copyright notice are from the [Boost library]{@link http://www.boost.org/doc/libs/1_64_0/boost/math/special_functions/factorials.hpp}. The implementation has been modified for JavaScript.
+*
+* ```text
+* (C) Copyright John Maddock 2006, 2010.
+*
+* Use, modification and distribution are subject to the
+* Boost Software License, Version 1.0. (See accompanying file
+* LICENSE or copy at http://www.boost.org/LICENSE_1_0.txt)
+* ```
+*/
````
```suggestion
```

This is not needed in `native.js` as there isn't anything derived from the Boost library.

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/rising-factorial/lib/native.js:
````diff
@@ -0,0 +1,85 @@
+
+/**
+* Computes the rising factorial of `x` and `n`.
+*
+* ## Notes
+*
+* -   The rising factorial is defined as
+*
+*     ```tex
+*     \operatorname{risingFactorial}(x, n) = x (x-1) (x-2) (x-3) \ldots (x-n+1)
+*     ```
+*
+*     or equivalently
+*
+*     ```tex
+*     \operatorname{risingFactorial}(x, n) = \frac{ \Gamma(x + n) }{ \Gamma(x) };
+*     ```
+*
````
```suggestion
```

anandkaranubc replied:
```suggestion
* @private
```

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/rising-factorial/manifest.json:
```diff
@@ -0,0 +1,81 @@
+      "libraries": [],
+      "libpath": [],
+      "dependencies": [
+        "@stdlib/math/base/assert/is-integer",
```
```suggestion
```

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/rising-factorial/benchmark/c/native/benchmark.c:
```diff
@@ -0,0 +1,138 @@
+
+	t = tic();
+	for ( i = 0; i < ITERATIONS; i++ ) {
+		y = stdlib_base_rising_factorial( x[ i % 100 ], n[ i % 100 ] );
```
```suggestion
		y = stdlib_base_rising_factorial( x[ i%100 ], n[ i%100 ] );
```

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/rising-factorial/benchmark/c/native/benchmark.c:
```diff
@@ -0,0 +1,138 @@
+
+	for ( i = 0; i < 100; i++ ) {
+		x[ i ] = ( 40.0 * rand_double() ) + 1.0;
+		n[ i ] = (int32_t)( 41.0 * rand_double() );
```
```suggestion
		x[ i ] = ( 100.0*rand_double() ) - 50.0;
```

Using similar values used in cpp benchmarks, for consistency.

anandkaranubc replied:
```suggestion
		n[ i ] = (int32_t)( 50.0*rand_double() );
```

Same comment.

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/rising-factorial/benchmark/benchmark.native.js:
```diff
@@ -0,0 +1,64 @@
+
+// MAIN //
+
+bench( pkg, opts, function benchmark( b ) {
```
```suggestion
bench( pkg+'::native', opts, function benchmark( b ) {
```

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/rising-factorial/benchmark/benchmark.native.js:
```diff
@@ -0,0 +1,64 @@
+	var i;
+
+	x = randu( 100, 1.0, 41.0 );
+	n = discreteUniform( 100, 0, 40 );
```
```suggestion
	x = uniform( 100, -50.0, 50.0 );
```

anandkaranubc replied:
```suggestion
	n = discreteUniform( 100, -50, 50 );
```

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/rising-factorial/benchmark/benchmark.native.js:
```diff
@@ -0,0 +1,64 @@
+
+	b.tic();
+	for ( i = 0; i < b.iterations; i++ ) {
+		y = risingFactorial( x[ i % x.length ], n[ i % n.length ] );
```
```suggestion
		y = risingFactorial( x[ i%x.length ], n[ i%n.length ] );
```

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/rising-factorial/examples/c/example.c:
```diff
@@ -0,0 +1,33 @@
+	int i;
+	for ( i = 0; i < 10; i++ ) {
+		v = stdlib_base_rising_factorial( x[ i ], n[ i ] );
+		printf( "x: %lf, n: %d, rising_factorial(x,n): %lf\n", x[ i ], n[ i ], v );
```
```suggestion
		printf( "risingFactorial(%lf, %d) = %lf\n", x[ i ], n[ i ], v );
```

Following the JS equivalent example file.

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/rising-factorial/benchmark/c/native/benchmark.c:
```diff
@@ -0,0 +1,138 @@
+*/
+
+#include "stdlib/math/base/special/rising_factorial.h"
+#include <stdlib.h>
```
```suggestion
#include <stdint.h>
```

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/rising-factorial/examples/index.js:
This file needs to be revisited using `logEachMap`

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/rising-factorial/test/test.native.js:
```diff
@@ -0,0 +1,105 @@
+
+tape( 'the function returns `1` if provided `n = 0` and a nonnegative `x`', opts, function test( t ) {
+	var val = risingFactorial( 2.0, 0 );
+	t.equal( val, 1.0, 'returns expected value' );
```
```suggestion
	t.strictEqual( val, 1.0, 'returns expected value' );
```

Applies here and elsewhere.

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/rising-factorial/README.md:
```diff
@@ -100,6 +100,86 @@ for ( i = 0; i < 100; i++ ) {
+
+#### stdlib_base_rising_factorial( x, n )
+
+Evaluates the rising factorial of `x` and `n`.
```
```suggestion
Evaluates the [rising factorial][falling-and-rising-factorials] of `x` and `n`.
```

anandkaranubc committed:
chore: apply suggestions from review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter commented on lib/node_modules/@stdlib/math/base/special/rising-factorial/src/main.c:
````diff
@@ -0,0 +1,116 @@
+* ```
+*/
+
+#include "stdlib/math/base/assert/is_integer.h"
````
Unused include:

```suggestion
```

Planeshifter commented on lib/node_modules/@stdlib/math/base/special/rising-factorial/src/main.c:
```diff
@@ -0,0 +1,116 @@
+* @param n     second function parameter
+* @return      function value
+*
+* @example
+* double v = stdlib_base_rising_factorial( 0.9, 5 );
+* // returns ~94.766
+*
+* @example
+* double v = stdlib_base_rising_factorial( -9.0, 3 );
+* // returns -504.0
+*
+* @example
+* double v = stdlib_base_rising_factorial( 0.0, 2 );
+* // returns 0.0
+*
+* @example
+* double v = stdlib_base_rising_factorial( 3.0, -2 );
+* // returns 0.5
```
```suggestion
* @example
* double v = stdlib_base_rising_factorial( 0.9, 5 );
* // returns ~94.766
```

Planeshifter committed:
chore: apply suggestions from code review

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>
